### PR TITLE
[zscaler_zpa] Add ZScaler ZPA package.

### PR DESCRIPTION
**Pull request by vinit-elastic**
## What does this PR do?

- Generated the skeleton of the ZScaler ZPA integration package.
- Added 5 data streams (App connector status, Audit, Browser access, User activity. and User status).
- Added data collection logic for all 5 data streams.
- Added the ingest pipeline for all the data streams.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files
- Added dashboards and visualizations.
- Added test for pipeline for all the data streams.
- Added system test cases for all the data streams.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.16.2`).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/zscaler_zpa directory.
- Run the following command to run tests. 
> elastic-package test 

## Screenshots

image
image
image
image

### CODE REVIEW

File: packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-app-connector-status.log-expected.json [hunk 1]
```diff
@@ -0,0 +1,107 @@
+{
+    "expected": [
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-app-connector-status.log-expected.json:2]: Can you please regenerate the expected files using the latest version of elastic-package. It sorts the keys which will make future changes easier to review.

File: packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-app-connector-status.log-expected.json [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                        "lon": -122,
+                        "lat": 47
+                    }
+                },
+                "os": {
+                    "platform": "el7"
+                },
+                "type": "forwarder",
+                "version": "19.20.3",
+                "ip": "0.0.0.0"
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-app-connector-status.log-expected.json:19]: Should be an array. https://www.elastic.co/guide/en/ecs/current/ecs-observer.html#field-observer-ip

File: packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-common-config.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+dynamic_fields:
+  event.ingested: ".*"
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-common-config.yml:2]: `event.ingested` isn't used in these tests so you can remove these 2 lines everywhere.

File: packages/zscaler_zpa/data_stream/app_connector_status/fields/ecs.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+- external: ecs
+  name: event.kind
+- external: ecs
+  name: event.type
+- external: ecs
+  name: host.cpu.usage
+- external: ecs
+  name: observer.geo.country_iso_code
+- description: Longitude and latitude
+  level: core
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/fields/ecs.yml:16]: level is not a field used in packages. https://github.com/elastic/package-spec/blob/aa2ec84dc6dcb041b5936cea3ef9e1c0106edc77/versions/1/data_stream/fields/fields.spec.yml#L10

File: packages/zscaler_zpa/data_stream/app_connector_status/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      fields:
+        - name: count
+          type: double
+          description: |
+            The number of services (combinations of domains/IP addresses and TCP/UDP ports) being monitored by the App Connector.
+    - name: primary_dns_resolver
+      type: keyword
+      description: |
+        The IP address of the primary DNS resolver.
+    - name: host_start_time
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/fields/fields.yml:67]: I wonder if we should make these IP fields as `type: ip` and for the times use `type: date`? The pipeline should have a `convert` processor to validate the IPs that's drops the field via `on_failure`. Same thing for the times.

File: packages/zscaler_zpa/data_stream/app_connector_status/manifest.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    title: Zscaler Private Access App Connector Status Logs
+    description: Collect Zscaler Private Access App Connector Status Logs using tcp input
+    vars:
+      - name: port
+        type: integer
+        title: TCP port to listen on
+        multi: false
+        required: true
+        show_user: true
+        default: 9005
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/manifest.yml:15]: I recommend greping through this repo to find out if this is the default for any other integrations.

File: packages/zscaler_zpa/data_stream/user_status/fields/fields.yml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+          type: keyword
+          description: |
+            The posture profiles that the Zscaler Client Connector verified for this device.
+        - name: miss
+          type: keyword
+          description: |
+            The posture profiles that the Zscaler Client Connector failed to verified for this device.
+    - name: private_ip
+      type: keyword
+      description: |
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/user_status/fields/fields.yml:36]: `type: ip` ?

**Review by andrewkroh — COMMENTED** [review]
Looks very good. Just a few minor suggestions.

File: packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-common-config.yml (hunk 3 above)
Comment (efd6) [reply packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-common-config.yml:2]: Also the cases where `event.ingested` is in sample events.

File: packages/zscaler_zpa/_dev/deploy/docker/sample_logs/app_connector_status.log [hunk 8]
```diff
@@ -0,0 +1 @@
+{"LogTimestamp":"Wed Jul 3 05:17:22 2019","Customer":"Customer Name","SessionID":"8A64Qwj9zCkfYDGJVoUZ","SessionType":"ZPN_ASSISTANT_BROKER_CONTROL","SessionStatus":"ZPN_STATUS_AUTHENTICATED","Version":"19.20.3","Platform":"el7","ZEN":"US-NY-8179","Connector":"Some App Connector","ConnectorGroup":"Some App Connector Group","PrivateIP":"10.0.0.4","PublicIP":"0.0.0.0","Latitude":47,"Longitude":-122,"CountryCode":"","TimestampAuthentication":"2019-06-27T05:05:23.348Z","TimestampUnAuthentication":"","CPUUtilization":1,"MemUtilization":20,"ServiceCount":2,"InterfaceDefRoute":"eth0","DefRouteGW":"10.0.0.1","PrimaryDNSResolver":"168.63.129.16","HostStartTime":"1513229995","HostUpTime":"1513229995","ConnectorUpTime":"1555920005","ConnectorStartTime":"1555920005","NumOfInterfaces":2,"BytesRxInterface":319831966346,"PacketsRxInterface":1617569938,"ErrorsRxInterface":0,"DiscardsRxInterface":0,"BytesTxInterface":192958782635,"PacketsTxInterface":1797471190,"ErrorsTxInterface":0,"DiscardsTxInterface":0,"TotalBytesRx":10902554,"TotalBytesTx":48931771}
```
Comment (efd6) [inline packages/zscaler_zpa/_dev/deploy/docker/sample_logs/app_connector_status.log:1]: Final newlines on non-generated files?

File: packages/zscaler_zpa/data_stream/app_connector_status/manifest.yml (hunk 6 above)
Comment (vinit-elastic) [reply packages/zscaler_zpa/data_stream/app_connector_status/manifest.yml:15]: I see the same port used for multiple integration packages. Is there any specific list of ports to use, as we have for geo-IP?

If we want to change it, then what should it be? 

cc: @efd6

File: packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-app-connector-status.log-expected.json (hunk 1 above)
Comment (vinit-elastic) [reply packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-app-connector-status.log-expected.json:2]: I have regenerated the test files.

File: packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-common-config.yml (hunk 3 above)
Comment (vinit-elastic) [reply packages/zscaler_zpa/data_stream/app_connector_status/_dev/test/pipeline/test-common-config.yml]: I have removed the section. 

And to @efd6's comment: We are not getting any cases where `event.ingested` is present.

File: packages/zscaler_zpa/_dev/deploy/docker/sample_logs/app_connector_status.log (hunk 8 above)
Comment (efd6) [reply packages/zscaler_zpa/_dev/deploy/docker/sample_logs/app_connector_status.log]: There are more of these in *.hbs and *.yml.

**Review by efd6 — APPROVED** [review]

File: packages/zscaler_zpa/manifest.yml [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  - src: /img/zscaler-logo.svg
+    title: Zscaler logo
+    size: 32x32
+    type: image/svg+xml
+policy_templates:
+  - name: zscaler_zpa
+    title: Zscaler Private Access logs
+    description: Collect Zscaler Private Access logs
+    inputs:
+      - type: tcp
```
Comment (andrewkroh) [inline packages/zscaler_zpa/manifest.yml:28]: Will you please rename this parameter to `listen_address`.

```
      - name: listen_address
        type: text
        title: Listen Address
        description: The bind address to listen for TCP connections. Set to `0.0.0.0` to bind to all available interfaces.
        multi: false
        required: true
        show_user: true
        default: localhost
```

File: packages/zscaler_zpa/data_stream/app_connector_status/manifest.yml [hunk 10]
```diff
@@ -0,0 +1,40 @@
+title: App Connector Status Logs
+type: logs
+streams:
+  - input: tcp
+    template_path: tcp.yml.hbs
+    title: Zscaler Private Access App Connector Status Logs
+    description: Collect Zscaler Private Access App Connector Status Logs using tcp input
+    vars:
+      - name: port
```
Comment (andrewkroh) [inline packages/zscaler_zpa/data_stream/app_connector_status/manifest.yml:9]: Will you please rename this to `listen_port` and do the same in the other manifests.

```yaml
      - name: listen_port
        type: integer
        title: Listen Port
        description: The TCP port number to listen on.
        multi: false
        required: true
        show_user: true
        default: 9010
```

**Pull request merged**